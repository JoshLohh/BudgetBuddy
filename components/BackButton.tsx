import React from 'react';
import { Pressable } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useRouter } from 'expo-router';

export default function BackButton({ color = 'black', size = 24, style = {} }) {
  const router = useRouter();
  return (
    <Pressable onPress={() => router.back()} style={[{ padding: 10 }, style]}>
      <Ionicons name="arrow-back" size={size} color={color} />
    </Pressable>
  );
}
