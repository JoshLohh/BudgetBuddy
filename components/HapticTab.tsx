import { Platform } from 'react-native';
import { BottomTabBarButtonProps } from '@react-navigation/bottom-tabs';
import { PlatformPressable } from '@react-navigation/elements';
import * as Haptics from 'expo-haptics';

export function HapticTab(props: BottomTabBarButtonProps) {
  return (
    <PlatformPressable
      {...props}
      onPressIn={ev => {
        if (Platform.OS === 'ios') {
          Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
        }
        props.onPressIn?.(ev);
      }}
      testID={props.testID}
    />
  );
}
