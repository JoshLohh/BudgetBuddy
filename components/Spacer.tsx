import { View, StyleSheet } from 'react-native';
import React from 'react';

type Dimension = number | `${number}%`;

type SpacerProps = {
  width?: Dimension;
  height?: Dimension;
  testID?: 'spacer';
};

const Spacer = ({ width = '100%', height = 40, testID }: SpacerProps) => {
  return <View style={{ width, height }} testID = {testID} />;
};

export default Spacer;


const styles = StyleSheet.create({})