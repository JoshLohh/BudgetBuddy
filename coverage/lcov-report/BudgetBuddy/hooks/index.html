
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for BudgetBuddy/hooks</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> BudgetBuddy/hooks</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">73.78% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>152/206</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">52.32% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>45/86</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.79% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>40/59</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">76.31% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>145/190</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="settlementUtils.ts"><a href="settlementUtils.ts.html">settlementUtils.ts</a></td>
	<td data-value="88.23" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 88%"></div><div class="cover-empty" style="width: 12%"></div></div>
	</td>
	<td data-value="88.23" class="pct high">88.23%</td>
	<td data-value="51" class="abs high">45/51</td>
	<td data-value="52.63" class="pct medium">52.63%</td>
	<td data-value="19" class="abs medium">10/19</td>
	<td data-value="78.57" class="pct medium">78.57%</td>
	<td data-value="14" class="abs medium">11/14</td>
	<td data-value="90.9" class="pct high">90.9%</td>
	<td data-value="44" class="abs high">40/44</td>
	</tr>

<tr>
	<td class="file empty" data-value="useColorScheme.ts"><a href="useColorScheme.ts.html">useColorScheme.ts</a></td>
	<td data-value="0" class="pic empty">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file low" data-value="useColorScheme.web.ts"><a href="useColorScheme.web.ts.html">useColorScheme.web.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	</tr>

<tr>
	<td class="file medium" data-value="useGroupDetails.ts"><a href="useGroupDetails.ts.html">useGroupDetails.ts</a></td>
	<td data-value="79.25" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 79%"></div><div class="cover-empty" style="width: 21%"></div></div>
	</td>
	<td data-value="79.25" class="pct medium">79.25%</td>
	<td data-value="135" class="abs medium">107/135</td>
	<td data-value="61.4" class="pct medium">61.4%</td>
	<td data-value="57" class="abs medium">35/57</td>
	<td data-value="72.5" class="pct medium">72.5%</td>
	<td data-value="40" class="abs medium">29/40</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="126" class="abs high">105/126</td>
	</tr>

<tr>
	<td class="file low" data-value="useGroups.ts"><a href="useGroups.ts.html">useGroups.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	</tr>

<tr>
	<td class="file low" data-value="useThemeColor.ts"><a href="useThemeColor.ts.html">useThemeColor.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	</tr>

<tr>
	<td class="file low" data-value="useUser.ts"><a href="useUser.ts.html">useUser.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-07-28T04:50:26.944Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    