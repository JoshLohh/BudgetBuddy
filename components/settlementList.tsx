import React, { useState } from 'react';
import { View, Alert } from 'react-native';
import { ThemedText } from '@/components/ThemedText';
import { ThemedButton } from '@/components/ThemedButton';
import { Ionicons } from '@expo/vector-icons';
import { Settlement } from '@/types';

interface SettlementListProps {
  suggestedSettlements: Settlement[];
  getUsername: (userId: string) => string;
  settleUp: (from: string, to: string, amount: number) => Promise<void>;
  currentUserId: string;
}

export default function SettlementList({
  suggestedSettlements,
  getUsername,
  settleUp,
  currentUserId,
}: SettlementListProps) {
  
  // Filter settlements for this user
  const userSettlements = suggestedSettlements.filter(
    s => s.from === currentUserId || s.to === currentUserId
  );

  // Calculate net balance
  let netBalance = 0;
  suggestedSettlements.forEach(s => {
    if (s.to === currentUserId) netBalance += s.amount;
    else if (s.from === currentUserId) netBalance -= s.amount;
  });

  // Helper for owed/owing statement and color
  let statement = '';
  let statementColor = '#888';
  if (userSettlements.length === 0) {
    statement = 'No settlements needed.';
  } else if (netBalance > 0) {
    statement = `You are owed $${netBalance.toFixed(2)}`;
    statementColor = '#00cc66';
  } else if (netBalance < 0) {
    statement = `You owe $${Math.abs(netBalance).toFixed(2)}`;
    statementColor = '#ff5050';
  } else {
    statement = 'No settlements needed.';
  }

  const [processing, setProcessing] = useState('');

  const handleSettleUp = (from: string, to: string, amount: number) => {
    Alert.alert(
      'Confirm Settle Up',
      `Are you sure you want to settle $${amount.toFixed(2)} from ${getUsername(from)} to ${getUsername(to)}?`,
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Confirm',
          onPress: async () => {
            setProcessing(`${from}_${to}_${amount}`);
            await settleUp(from, to, amount);
            setProcessing('');
          },
        },
      ]
    );
  };

  return (
    <View style={{ marginTop: 12 }}>
      <ThemedText type="subtitle" style={{ fontWeight: 'bold', marginBottom: 8 }}>
        Suggested Settlements
      </ThemedText>
      <ThemedText style={{ fontWeight:'bold', color: statementColor }}>
        {statement}
      </ThemedText>
      {userSettlements.length === 0 ? (
        <></>
      ) : (
        userSettlements.map(({ from, to, amount }) => (
          <View
            key={`${from}_${to}_${amount}`}
            style={{ flexDirection: 'row', alignItems: 'center', marginBottom: 8 }}
          >
            <ThemedText style={{ flex: 1 , fontWeight:'bold'}}>
              {getUsername(from)} <Ionicons name="arrow-forward" size={16} color="#1976d2" /> {getUsername(to)}
            </ThemedText>
            <ThemedText style={{ width: 80 , color: '#1e88e5', fontWeight: 'bold'}}>${amount.toFixed(2)}</ThemedText>
            <ThemedButton
              onPress={() => handleSettleUp(from, to, amount)}
              disabled={processing === `${from}_${to}_${amount}`}
              style={{ marginLeft: 10, minWidth: 90 }}
            >
              <ThemedText style={{ color: '#fff' }}>Settle Up</ThemedText>
            </ThemedButton>
          </View>
        ))
      )} 
    </View>
  );
}
