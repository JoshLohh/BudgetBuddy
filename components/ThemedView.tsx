import { View, type ViewProps } from 'react-native';

import { useThemeColor } from '@/hooks/useThemeColor';
import { SafeAreaView, useSafeAreaInsets } from 'react-native-safe-area-context';

export type ThemedViewProps = ViewProps & {
  lightColor?: string;
  darkColor?: string;
  safe?: boolean;
  testID?: string;
};

export function ThemedView({ style, safe = false,  lightColor, darkColor, ...otherProps }: ThemedViewProps,) {
  const backgroundColor = useThemeColor({ light: lightColor, dark: darkColor }, 'background');

  if (!safe) return (
  <View testID="themed-view" style={[{ backgroundColor }, style]} {...otherProps} />
  )

  const insets = useSafeAreaInsets()

  return (
    <View 
    testID="themed-view"
    style={[{
      backgroundColor,
      paddingTop: insets.top,
      paddingBottom: insets.bottom,
      }, style]} 
       {...otherProps} />
  )
}
