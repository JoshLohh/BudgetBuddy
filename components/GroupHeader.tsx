import React, { useState, useEffect } from 'react';
import { View, StyleSheet, ActivityIndicator } from 'react-native';
import { ThemedView } from '@/components/ThemedView';
import { ThemedText } from '@/components/ThemedText';
import ThemedTextInput from '@/components/ThemedTextInput';
import { ThemedButton } from '@/components/ThemedButton';
import { Ionicons } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { databases } from '@/lib/appwrite';
import { router } from 'expo-router';

const databaseId = process.env.EXPO_PUBLIC_APPWRITE_DATABASE_ID ?? '';
const groupsCollectionId = process.env.EXPO_PUBLIC_APPWRITE_GROUPS_COLLECTION_ID ?? '';

export default function GroupHeader({ group, totalExpenses, onGroupUpdated }) {
  const [editing, setEditing] = useState(false);
  const [title, setTitle] = useState(group?.title || '');
  const [description, setDescription] = useState(group?.description || '');
  const [saving, setSaving] = useState(false);
  const [error, setError] = useState('');

  useEffect(() => {
    setTitle(group?.title || '');
    setDescription(group?.description || '');
  }, [group]);

  if (!group) {
    return (
      <View style={[styles.container, { alignItems: 'center', justifyContent: 'center', minHeight: 80 }]}>
        <ActivityIndicator />
      </View>
    );
  }

  const handleSave = async () => {
    setSaving(true);
    setError('');
    try {
      await databases.updateDocument(databaseId, groupsCollectionId, group.id, {
        title,
        description,
      });
      setEditing(false);
      if (onGroupUpdated) {
        onGroupUpdated({ ...group, title, description });
      }
    } catch (e) {
      setError('Failed to update group details.');
    }
    setSaving(false);
  };

  if (editing) {
    return (
      <ThemedView style={styles.container}>
        <View style={styles.topRow}>
          <View style={styles.leftCol}>
            <ThemedTextInput
              label="Group Name"
              value={title}
              onChangeText={setTitle}
              style={{ marginBottom: 8 }}
            />
            <ThemedTextInput
              label="Description"
              value={description}
              onChangeText={setDescription}
              style={{ marginBottom: 8 }}
            />
            {error ? (
              <ThemedText style={{ color: 'red', marginBottom: 8 }}>{error}</ThemedText>
            ) : null}
          </View>
          <View style={styles.rightCol}>
            <View style={styles.totalCard}>
              <ThemedText style={styles.totalLabel}>Total</ThemedText>
              <ThemedText style={styles.totalAmount}>${totalExpenses.toFixed(2)}</ThemedText>
            </View>
            <ThemedButton
              onPress={handleSave}
              disabled={saving}
              style={styles.editBtn}
            >
              <ThemedText style={{ color: '#fff' }}>
                {saving ? 'Saving...' : 'Save'}
              </ThemedText>
            </ThemedButton>
            <ThemedButton
              onPress={() => setEditing(false)}
              style={[styles.editBtn, { backgroundColor: '#e0e0e0', marginTop: 8 }]}
            >
              <ThemedText style={{ color: '#333' }}>Cancel</ThemedText>
            </ThemedButton>
          </View>
        </View>
      </ThemedView>
    );
  }

  return (
    <ThemedView style={styles.container}>
      <View style={styles.topRow}>
        <View style={styles.leftCol}>
          <ThemedText type="title" style={styles.groupTitle}>{group.title}</ThemedText>
          {group.description ? (
            <ThemedText style={styles.groupDescription}>{group.description}</ThemedText>
          ) : null}
          <ThemedText style={styles.membersCount}>
            {group.members?.length ?? 0} member{(group.members?.length ?? 0) !== 1 ? 's' : ''}
          </ThemedText>
          <ThemedButton
            onPress={() => router.push({ pathname: '/group/[groupId]/history', params: { groupId: group.id } })}
            style={{
              marginTop: 4,           
              marginBottom: 4,        
              height: 32,             
              alignSelf: 'flex-start',    
              justifyContent: 'center',
              alignItems: 'center',
              paddingVertical: 0,
              paddingHorizontal: 14,
              minWidth: 200,
            }}
          >
            <ThemedText style={{ textAlign: 'center', fontSize: 12, color:'#fff' }}>View Activity Log</ThemedText>
          </ThemedButton>
        </View>
        <View style={styles.rightCol}>
          <View style={styles.totalCard}>
            <ThemedText style={styles.totalLabel}>Total</ThemedText>
            <ThemedText style={styles.totalAmount}>${totalExpenses.toFixed(2)}</ThemedText>
          </View>
          <ThemedButton
            onPress={() => setEditing(true)}
            style={styles.editBtn}
          >
            <Ionicons name="pencil" size={18} color="#fff" />
          </ThemedButton>
        </View>
      </View>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#fff', 
    padding: 10,
    marginBottom: 10,
    borderRadius: 12,
  },
  topRow: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
  },
  leftCol: {
    flex: 1,
    marginRight: 12,
  },
  rightCol: {
    alignItems: 'flex-end',
    minWidth: 110,
  },
  groupTitle: {
    fontSize: 28,
    fontWeight: '700',
    color: 'black'
  },
  groupDescription: {
    fontSize: 15,
    color: '#555',
    marginBottom: 4,
  },
  membersCount: {
    fontSize: 14,
    color: '#888',
    marginBottom: 8,
  },
  totalCard: {
    backgroundColor: '#b3e0ff', 
    borderRadius: 10,
    paddingVertical: 8,
    paddingHorizontal: 18,
    alignItems: 'center',
    minWidth: 110,
    elevation: 2,
    marginBottom: 10,
    alignSelf: 'flex-end',
  },
  totalLabel: {
    fontSize: 13,
    color: '#1976d2',
    fontWeight: 'bold',
  },
  totalAmount: {
    fontSize: 18,
    color: '#1976d2',
    fontWeight: 'bold',
    marginTop: 2,
  },
  editBtn: {
    backgroundColor: Colors.primary,
    borderRadius: 8,
    paddingVertical: 10,
    paddingHorizontal: 0,
    width: 110,
    alignSelf: 'flex-end',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
