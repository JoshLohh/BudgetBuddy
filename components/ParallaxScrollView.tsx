import type { PropsWithChildren, ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import Animated, {
  interpolate,
  useAnimatedRef,
  useAnimatedStyle,
  useScrollViewOffset,
} from 'react-native-reanimated';

import { ThemedView } from '@/components/ThemedView';
import { useBottomTabOverflow } from '@/components/ui/TabBarBackground';
import { useColorScheme } from '@/hooks/useColorScheme';

const HEADER_HEIGHT = 250;

type Props = PropsWithChildren<{
  headerImage?: ReactElement;
  headerBackgroundColor?: { dark: string; light: string };
}>;

export default function ParallaxScrollView({
  children = null,
  headerImage,
  headerBackgroundColor = { dark: '#222', light: '#fff' },
}: Props) {
  const colorScheme = useColorScheme() ?? 'light';

  const scrollRef = useAnimatedRef<Animated.ScrollView>();
  const scrollOffset = useScrollViewOffset(scrollRef);

  const bottom = useBottomTabOverflow();

  // Animate the header image for parallax effect
  const headerAnimatedStyle = useAnimatedStyle(() => ({
    transform: [
      {
        translateY: interpolate(
          scrollOffset.value,
          [-HEADER_HEIGHT, 0, HEADER_HEIGHT],
          [-HEADER_HEIGHT / 2, 0, HEADER_HEIGHT * 0.75]
        ),
      },
      {
        scale: interpolate(
          scrollOffset.value,
          [-HEADER_HEIGHT, 0, HEADER_HEIGHT],
          [2, 1, 1]
        ),
      },
    ],
  }));

  return (
    <ThemedView style={styles.container}>
      {/* Header Image */}
      <View style={[styles.header, { backgroundColor: headerBackgroundColor[colorScheme] }]}>
        {headerImage ? (
          <Animated.View style={[StyleSheet.absoluteFill, headerAnimatedStyle]}>
            {headerImage}
          </Animated.View>
        ) : null}
      </View>
      {/* Content */}
      <Animated.ScrollView
        ref={scrollRef}
        contentContainerStyle={[
          styles.content,
          { paddingBottom: bottom + 32 },
        ]}
        showsVerticalScrollIndicator={false}
        bounces
      >
        {children}
      </Animated.ScrollView>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  header: {
    height: HEADER_HEIGHT,
    overflow: 'hidden',
    width: '100%',
  },
  content: {
    flexGrow: 1,
    paddingHorizontal: 24,
    paddingTop: 16,
    gap: 16,
    backgroundColor: 'transparent',
  },
});
