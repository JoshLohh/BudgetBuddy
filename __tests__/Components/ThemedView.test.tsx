import React from 'react';
import { render } from '@testing-library/react-native';
import * as ThemeHook from '@/hooks/useThemeColor';
import * as SafeArea from 'react-native-safe-area-context';
import { ThemedView } from '../../components/ThemedView';

// Mock SafeArea context
jest.mock('react-native-safe-area-context', () => {
  const actual = jest.requireActual('react-native-safe-area-context');

  return {
    ...actual,
    useSafeAreaInsets: jest.fn().mockReturnValue({ top: 10, bottom: 20, left: 0, right: 0 }),
  };
});

describe('ThemedView', () => {
  beforeEach(() => {
    jest.spyOn(ThemeHook, 'useThemeColor').mockImplementation(() => '#abcdef');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('renders a View when safe is false (default)', () => {
    const { getByTestId } = render(<ThemedView testID="themed-view" />);
    const view = getByTestId('themed-view');

    expect(view.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ backgroundColor: '#abcdef' }),
      ])
    );
  });

  it('renders a SafeAreaView when safe is true and respects insets', () => {
    const { getByTestId } = render(<ThemedView safe testID="safe-view" />);
    const safeView = getByTestId('safe-view');

    expect(safeView.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          backgroundColor: '#abcdef',
          paddingTop: 10,
          paddingBottom: 20,
        }),
      ])
    );
  });

  it('merges custom style with themed background', () => {
    const customStyle = { borderRadius: 8 };
    const { getByTestId } = render(
      <ThemedView style={customStyle} testID="styled-view" />
    );
    const view = getByTestId('styled-view');

    expect(view.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ borderRadius: 8 }),
        expect.objectContaining({ backgroundColor: '#abcdef' }),
      ])
    );
  });

  it('uses lightColor and darkColor props in useThemeColor call', () => {
    const spy = jest
      .spyOn(ThemeHook, 'useThemeColor')
      .mockImplementation(({ light, dark }) => light ?? '#123456');

    render(
      <ThemedView
        testID="view-with-custom-colors"
        lightColor="#aabbcc"
        darkColor="#ddeeff"
      />
    );

    expect(spy).toHaveBeenCalledWith(
      { light: '#aabbcc', dark: '#ddeeff' },
      'background'
    );
  });

  it('matches snapshot for View (safe = false)', () => {
    const tree = render(<ThemedView testID="snapshot-view" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot for SafeAreaView (safe = true)', () => {
    const tree = render(<ThemedView safe testID="snapshot-safe-view" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

/**
 * Test Summary:
 * - Confirms correct conditional rendering between View and SafeAreaView.
 * - Mocks useThemeColor and useSafeAreaInsets to isolate theming and layout logic.
 * - Tests merging of theme-derived and custom styles.
 * - Snapshot tests ensure output stability as component evolves.
 */
