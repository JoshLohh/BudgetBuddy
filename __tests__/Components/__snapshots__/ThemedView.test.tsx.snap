// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemedView matches snapshot for SafeAreaView (safe = true) 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#abcdef",
        "paddingBottom": 20,
        "paddingTop": 10,
      },
      undefined,
    ]
  }
  testID="snapshot-safe-view"
/>
`;

exports[`ThemedView matches snapshot for View (safe = false) 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#abcdef",
      },
      undefined,
    ]
  }
  testID="snapshot-view"
/>
`;
